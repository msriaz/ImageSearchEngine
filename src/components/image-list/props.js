import PropTypes from 'prop-types';

export const propTypes = {
  data: PropTypes.array.isRequired,
  loadMore: PropTypes.func.isRequired,
  isLoadingMore: PropTypes.bool.isRequired,
};

export const defaultProps = {};
