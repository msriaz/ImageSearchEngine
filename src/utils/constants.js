import {color as colors} from '../theme';
import {generateColorPalleteConstants} from './index';

export const APP_COLORS = generateColorPalleteConstants(colors);

export const STORAGE_KEYS = {
  HISTORY: 'history',
};
