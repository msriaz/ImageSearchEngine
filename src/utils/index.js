import {useEffect, useRef} from 'react';
import snakeCase from 'lodash/snakeCase';
/**
 *
 * @param {*} value
 * @returns previous state
 */
export const usePrevious = value => {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

/**
 *
 * @param {*} imgData
 * @returns image url to render
 */
export const getImageUrl = imgData => {
  const {farm, server, id, secret} = imgData;
  return `https://farm${farm}.static.flickr.com/${server}/${id}_${secret}.jpg`;
};

/**
 * generate color consts from color pallette
 * @param {*} colors
 */
export const generateColorPalleteConstants = colors => {
  const colorsList = {};
  Object.keys(colors).forEach(colorKey => {
    colorsList[snakeCase(colorKey).toUpperCase()] = colors[colorKey];
  });
  return colorsList;
};

/**
 *
 * @param {*} funcName
 * @param {*} error
 */
export const reportError = (funcName, error) => {
  console.log(`Error in ${funcName}`, error);
};
