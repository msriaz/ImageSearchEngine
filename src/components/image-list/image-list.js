import React, {forwardRef} from 'react';
import {FlatList} from 'react-native';
import {props, defaultProps} from './props';

import {ImageItem} from './image-item';
import {Spinner} from '../spinner';

/**
 * Image Item Component
 * @param {data} defines the data of images to render
 * @param {loadMore} defines the function to be called for loading more data
 * @param {rest} defines additional props to pass to FlatList
 * @param {isLoadingMore} defines to load more
 */
export const ImageList = forwardRef(
  ({data, loadMore, isLoadingMore, rest}, ref) => {
    const renderItem = ({item}) => <ImageItem imageData={item} key={item.id} />;

    const ListFooterComponent = () =>
      isLoadingMore ? <Spinner variant={Spinner.VARIANTS.SMALL} /> : null;

    return (
      <FlatList
        ref={ref}
        data={data}
        renderItem={renderItem}
        horizontal={false}
        numColumns={2}
        onEndReached={loadMore}
        ListFooterComponent={ListFooterComponent}
        {...rest}
      />
    );
  },
);

ImageList.props = props;
ImageList.defaultProps = defaultProps;
