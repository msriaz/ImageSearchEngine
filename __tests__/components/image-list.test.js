import React from 'react';
import {ImageList} from '../../src/components/image-list';
import {render} from '../utils/theme-wrapper';

const onPressMock = jest.fn();

const mockData = [
  {
    id: '51443585974',
    secret: '0e481ab487',
    server: '65535',
    farm: 66,
  },
  {
    id: '51442086532',
    secret: '39646048b3',
    server: '65535',
    farm: 66,
  },
  {
    id: '51443091678',
    secret: '6e5837e79a',
    server: '65535',
    farm: 66,
  },
];

describe('Image List Component', () => {
  it('should match the Image List  snapshot', () => {
    const compTree = render(
      <ImageList data={mockData} loadMore={onPressMock} isLoadingMore />,
    );
    expect(compTree.toJSON()).toMatchSnapshot();
  });
});
