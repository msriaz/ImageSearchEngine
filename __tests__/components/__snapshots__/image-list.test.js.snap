// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image List Component should match the Image List  snapshot 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "farm": 66,
        "id": "51443585974",
        "secret": "0e481ab487",
        "server": "65535",
      },
      Object {
        "farm": 66,
        "id": "51442086532",
        "secret": "39646048b3",
        "server": "65535",
      },
      Object {
        "farm": 66,
        "id": "51443091678",
        "secret": "6e5837e79a",
        "server": "65535",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[MockFunction]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "height": "NaNpx",
                "position": "absolute",
                "width": "45%",
              },
            ]
          }
        >
          <ActivityIndicator
            color="#A8A8A8"
            size="small"
            style={
              Array [
                Object {
                  "left": "50%",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "top": "50%",
                  "zIndex": 2,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F4F4F4",
                  "height": "NaNpx",
                  "justifyContent": "center",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "width": "45%",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": "NaNpx",
                      "marginBottom": 8,
                      "marginLeft": 8,
                      "marginRight": 8,
                      "marginTop": 8,
                      "resizeMode": "cover",
                      "width": "45%",
                    },
                    Object {
                      "position": "absolute",
                    },
                  ],
                ]
              }
            >
              <FastImageView
                resizeMode="cover"
                source={
                  Object {
                    "uri": "https://farm66.static.flickr.com/65535/51443585974_0e481ab487.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "height": "NaNpx",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "resizeMode": "cover",
                  "width": "45%",
                },
              ],
            ]
          }
        >
          <FastImageView
            onFastImageLoad={[Function]}
            onFastImageLoadStart={[Function]}
            resizeMode="cover"
            source={
              Object {
                "uri": "https://farm66.static.flickr.com/65535/51443585974_0e481ab487.jpg",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "height": "NaNpx",
                "position": "absolute",
                "width": "45%",
              },
            ]
          }
        >
          <ActivityIndicator
            color="#A8A8A8"
            size="small"
            style={
              Array [
                Object {
                  "left": "50%",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "top": "50%",
                  "zIndex": 2,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F4F4F4",
                  "height": "NaNpx",
                  "justifyContent": "center",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "width": "45%",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": "NaNpx",
                      "marginBottom": 8,
                      "marginLeft": 8,
                      "marginRight": 8,
                      "marginTop": 8,
                      "resizeMode": "cover",
                      "width": "45%",
                    },
                    Object {
                      "position": "absolute",
                    },
                  ],
                ]
              }
            >
              <FastImageView
                resizeMode="cover"
                source={
                  Object {
                    "uri": "https://farm66.static.flickr.com/65535/51442086532_39646048b3.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "height": "NaNpx",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "resizeMode": "cover",
                  "width": "45%",
                },
              ],
            ]
          }
        >
          <FastImageView
            onFastImageLoad={[Function]}
            onFastImageLoadStart={[Function]}
            resizeMode="cover"
            source={
              Object {
                "uri": "https://farm66.static.flickr.com/65535/51442086532_39646048b3.jpg",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "height": "NaNpx",
                "position": "absolute",
                "width": "45%",
              },
            ]
          }
        >
          <ActivityIndicator
            color="#A8A8A8"
            size="small"
            style={
              Array [
                Object {
                  "left": "50%",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "top": "50%",
                  "zIndex": 2,
                },
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F4F4F4",
                  "height": "NaNpx",
                  "justifyContent": "center",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "position": "absolute",
                  "width": "45%",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "height": "NaNpx",
                      "marginBottom": 8,
                      "marginLeft": 8,
                      "marginRight": 8,
                      "marginTop": 8,
                      "resizeMode": "cover",
                      "width": "45%",
                    },
                    Object {
                      "position": "absolute",
                    },
                  ],
                ]
              }
            >
              <FastImageView
                resizeMode="cover"
                source={
                  Object {
                    "uri": "https://farm66.static.flickr.com/65535/51443091678_6e5837e79a.jpg",
                  }
                }
                style={
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "height": "NaNpx",
                  "marginBottom": 8,
                  "marginLeft": 8,
                  "marginRight": 8,
                  "marginTop": 8,
                  "resizeMode": "cover",
                  "width": "45%",
                },
              ],
            ]
          }
        >
          <FastImageView
            onFastImageLoad={[Function]}
            onFastImageLoadStart={[Function]}
            resizeMode="cover"
            source={
              Object {
                "uri": "https://farm66.static.flickr.com/65535/51443091678_6e5837e79a.jpg",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <ActivityIndicator
        size="small"
      />
    </View>
  </View>
</RCTScrollView>
`;
